discovery-provider/src/tasks/index.py

```diff
@@ -133,6 +134,28 @@ def fetch_tx_receipts(self, block_transactions):
         raise Exception(f"index.py | fetch_tx_receipts Expected ${num_submitted_txs} received {num_processed_txs}")
     return block_tx_with_receipts
 
+# During each indexing iteration, check if the address for UserReplicaSetManager
+# has been set in the L2 contract registry - if so, update the global contract_addresses object
+# This change is to ensure no indexing restart is necessary when UserReplicaSetManager is
+# added to the registry.
+def update_user_replica_set_manager_address_if_necessary(self):
+    web3 = update_task.web3
+    shared_config = update_task.shared_config
+    abi_values = update_task.abi_values
+    user_replica_set_manager_address = contract_addresses["user_replica_set_manager"]
+    if user_replica_set_manager_address == "0x0000000000000000000000000000000000000000":
+        registry_address = web3.toChecksumAddress(
+            shared_config["contracts"]["registry"]
+        )
+        registry_instance = web3.eth.contract(
+            address=registry_address, abi=abi_values["Registry"]["abi"]
+        )
+        user_replica_set_manager_address = registry_instance.functions.getContract(
+            bytes("UserReplicaSetManager", "utf-8")
+        ).call()
+        contract_addresses["user_replica_set_manager"] = web3.toChecksumAddress(user_replica_set_manager_address)
```

Comment:
Not sure off the top of my head how setting a global variable like this across files in python works but i trust you tested this and it works?

Reply:
yes

Reply:
Confirmation of dynamically updating disc prov addresses (note timestamps):

```
Logs prior to deployment of UserReplicaSetManager contract:
{"levelno": 20, "level": "INFO", "msg": "index.py | No update to user_replica_set_manager address, queried 0x0000000000000000000000000000000000000000 from registry", "timestamp": "2021-02-09 22:27:52,416"}
{"levelno": 20, "level": "INFO", "msg": "index.py | get_latest_block | current=31 target=31", "timestamp": "2021-02-09 22:27:52,434"}
{"levelno": 20, "level": "INFO", "msg": "index.py | update_task | Intersect_blockhash : 0x6b3abad9d4d22450ae2269413e6ed139d0aa53615ced75b59e8a77a5a7e641d0", "timestamp": "2021-02-09 22:27:52,452"}
{"levelno": 20, "level": "INFO", "msg": "index.py | update_task | 55e3a4fc-7e37-4c2f-991a-7d6d593f4390 | Processing complete within session", "timestamp": "2021-02-09 22:27:52,453"}
{"levelno": 20, "level": "INFO", "msg": "Task update_discovery_provider[55e3a4fc-7e37-4c2f-991a-7d6d593f4390] succeeded in 0.13760744407773018s: None", "timestamp": "2021-02-09 22:27:52,453", "data": {"id": "55e3a4fc-7e37-4c2f-991a-7d6d593f4390", "name": "update_discovery_provider", "return_value": "None", "runtime": 0.13760744407773018}}
{"levelno": 20, "level": "INFO", "msg": "Received task: update_discovery_provider[ccc7d1a5-c903-4c0c-8687-6c6e30649324]  ", "timestamp": "2021-02-09 22:27:57,313"}
{"levelno": 20, "level": "INFO", "msg": "Received task: update_ipld_blacklist[e44cfb7b-eba0-4828-bc28-1e30137629b5]  ", "timestamp": "2021-02-09 22:27:57,315"}
{"levelno": 20, "level": "INFO", "msg": "index.py | ccc7d1a5-c903-4c0c-8687-6c6e30649324 | update_task | Acquired disc_prov_lock", "timestamp": "2021-02-09 22:27:57,326"}
{"levelno": 20, "level": "INFO", "msg": "IPLDBLACKLIST | get_latest_blacklist_block | current=31 target=31", "timestamp": "2021-02-09 22:27:57,347"}
{"levelno": 20, "level": "INFO", "msg": "IPLDBLACKLIST | Exited block intersection loop - True", "timestamp": "2021-02-09 22:27:57,363"}
{"levelno": 20, "level": "INFO", "msg": "Task update_ipld_blacklist[gge44cfb7b-eba0-4828-bc28-1e30137629b5] succeeded in 0.0466228099539876s: None", "timestamp": "2021-02-09 22:27:57,364", "data": {"id": "e44cfb7b-eba0-4828-bc28-1e30137629b5", "name": "update_ipld_blacklist", "return_value": "None", "runtime": 0.0466228099539876}}
{"levelno": 20, "level": "INFO", "msg": "index.py | No update to user_replica_set_manager address, queried 0x0000000000000000000000000000000000000000 from registry", "timestamp": "2021-02-09 22:27:57,388"}
...
{"levelno": 20, "level": "INFO", "msg": "index.py | No update to user_replica_set_manager address, queried 0x0000000000000000000000000000000000000000 from registry", "timestamp": "2021-02-09 22:30:07,459"}
{"levelno": 20, "level": "INFO", "msg": "index.py | get_latest_block | current=31 target=31", "timestamp": "2021-02-09 22:30:07,472"}
```
Log of restart action (note timestamp):
```
ubuntu@ip-172-31-29-163:~/audius-protocol/service-commands/scripts$ date -d @1234567890 && node setup.js run user-replica-set-manager up
Fri Feb 13 23:31:30 UTC 2009
Config dir: /home/ubuntu/.audius
/home/ubuntu/.audius/config.json
/home/ubuntu/.audius/eth-config.json
found service

...

   -------------------------------------
   > Total cost:           0.0749421 ETH


Summary
=======
> Total deployments:   2
> Final cost:          0.0749421 ETH


user-replica-set-manager - up | executed in 16.892s
ubuntu@ip-172-31-29-163:~/audius-protocol/service-commands/scripts$ date -d @1234567890
Fri Feb 13 23:31:30 UTC 2009
```

Logs immediately after deployment (no restart):
```
{"levelno": 20, "level": "INFO", "msg": "index.py | Updated user_replica_set_manager_address=0x8FCed9d95293d29064Bde46eaA39118813975e48", "timestamp": "2021-02-09 22:30:57,515"}
{"levelno": 20, "level": "INFO", "msg": "index.py | get_latest_block | current=33 target=36", "timestamp": "2021-02-09 22:30:57,535"}
{"levelno": 20, "level": "INFO", "msg": "index.py | update_task | Intersect_blockhash : 0x81e9112f1430785c230afd27067952330140ee765974092e9f72b12097808bb7", "timestamp": "2021-02-09 22:30:57,614"}
{"levelno": 20, "level": "INFO", "msg": "index.py | index_blocks | 93a42918-3e32-41d1-b37e-60b5f913d6d7 | block 34 - 1/3", "timestamp": "2021-02-09 22:30:57,615"}
{"levelno": 20, "level": "INFO", "msg": "index.py | User Replica Set Manager contract addr: 0x8FCed9d95293d29064Bde46eaA39118813975e48
...
{"levelno": 30, "level": "WARNING", "msg": "/usr/local/lib/python3.7/site-packages/web3/contract.py:1154: UserWarning: The log with transaction hash: b'\\x15\\x13\\x90\\xe9.\\xb1\\xdc\\x01a}\\xc2%y\\x92\\xcdl\\xa2\\xe1\\xe8\\x0f(\\xe8\\xfb?\\xdapQj\\x97\\xdf\\xe40' and logIndex: 0 encountered the following error during processing: MismatchedABI(The event signature did not match the provided ABI). It has been discarded.\n  f\"The log with transaction hash: {log['transactionHash']} and \"", "timestamp": "2021-02-09 22:30:57,658"}
{"levelno": 30, "level": "WARNING", "msg": "/usr/local/lib/python3.7/site-packages/web3/contract.py:1154: UserWarning: The log with transaction hash: b'\\x15\\x13\\x90\\xe9.\\xb1\\xdc\\x01a}\\xc2%y\\x92\\xcdl\\xa2\\xe1\\xe8\\x0f(\\xe8\\xfb?\\xdapQj\\x97\\xdf\\xe40' and logIndex: 1 encountered the following error during processing: MismatchedABI(The event signature did not match the provided ABI). It has been discarded.\n  f\"The log with transaction hash: {log['transactionHash']} and \"", "timestamp": "2021-02-09 22:30:57,659"}
{"levelno": 20, "level": "INFO", "msg": "3/3 blocks", "timestamp": "2021-02-09 22:30:57,660"}

{"levelno": 20, "level": "INFO", "msg": "Task update_ipld_blacklist[45cb48ca-9c82-480a-bffc-417d6aa487fb] succeeded in 0.2582509289495647s: None", "timestamp": "2021-02-09 22:30:57,698", "data": {"id": "45cb48ca-9c82-480a-bffc-417d6aa487fb", "name": "update_ipld_blacklist", "return_value": "None", "runtime": 0.2582509289495647}}
{"levelno": 20, "level": "INFO", "msg": "user_replica_set.py | Content Node Processing Adding <USRMContentNode(blockhash=0xfdd7c24a33d50db665e4311615e705169c3ce950a28f56ad6589456c4909f692,blocknumber=34,is_current=True,cnode_sp_id=1,delegate_owner_wallet=0x45781106706598C01a5D410599a7C8f966E6d384,proposer_sp_ids=[0, 0, 0],proposer_1_address=0x0000000000000000000000000000000000000000,proposer_2_address=0x0000000000000000000000000000000000000000,proposer_3_address=0x0000000000000000000000000000000000000000)>", "timestamp": "2021-02-09 22:30:57,703"}
{"levelno": 20, "level": "INFO", "msg": "user_replica_set.py | Content Node Processing Adding <USRMContentNode(blockhash=0xfdd7c24a33d50db665e4311615e705169c3ce950a28f56ad6589456c4909f692,blocknumber=34,is_current=True,cnode_sp_id=2,delegate_owner_wallet=0xC415A2241B50Db6bF8BAbBcE874edef4994c1e80,proposer_sp_ids=[0, 0, 0],proposer_1_address=0x0000000000000000000000000000000000000000,proposer_2_address=0x0000000000000000000000000000000000000000,proposer_3_address=0x0000000000000000000000000000000000000000)>", "timestamp": "2021-02-09 22:30:57,708"}
{"levelno": 20, "level": "INFO", "msg": "user_replica_set.py | Content Node Processing Adding <USRMContentNode(blockhash=0xfdd7c24a33d50db665e4311615e705169c3ce950a28f56ad6589456c4909f692,blocknumber=34,is_current=True,cnode_sp_id=3,delegate_owner_wallet=0x2A5AaaBC16982151021f7A42509D23d61B62a787,proposer_sp_ids=[0, 0, 0],proposer_1_address=0x0000000000000000000000000000000000000000,proposer_2_address=0x0000000000000000000000000000000000000000,proposer_3_address=0x0000000000000000000000000000000000000000)>", "timestamp": "2021-02-09 22:30:57,714"}
{"levelno": 20, "level": "INFO", "msg": "index.py | index_blocks | 93a42918-3e32-41d1-b37e-60b5f913d6d7 | block 35 - 2/3", "timestamp": "2021-02-09 22:30:57,721"}
```